src/util/schema_keywords.js

```diff
@@ -0,0 +1,174 @@
+            // idates in the database schema into the date format (currently most of them are
+            // already saved in ISO format which is not idate)
+            // TODO: Remove accepting instanceof Date after converting all uses of
+            // idates in the database schema into the date format (currently most of them are
+            // already saved in ISO format which is not idate)
+            const d = cxt.it.data;
+            cxt.gen
+                .if(CG`
+                    Number.isInteger(${d}) &&
+                    !isNaN(new Date(${d}).getTime())
+                `) // good idate
+                .elseIf(CG` ${d} instanceof Date `) // ok but see todo ^
+                .else()
+                .code(() => cxt.error())
+                .endIf();
+        }
+    },
+
+    // schema: { objectid: true } will match (new mongodb.ObjectId()) or (new mongodb.ObjectId()).valueOf()
+    objectid: {
+        keyword: 'objectid',
+        // schemaType: 'boolean',
+        /**
+         * 
+         * @param {KeywordCxt} cxt 
+         * 
+         */
+        code(cxt) {
+            const d = cxt.it.data;
+            cxt.gen
+                .if(CG`
+                    typeof ${d} === 'object' &&
+                    ${d} && 
+                    ${d}.constructor && 
+                    ${d}.constructor.name === 'ObjectID'
+                `)
+                .elseIf(CG`
+                    typeof ${d} === 'string' &&
+                    /^[0-9a-fA-F]{24}$/.test(${d})
+                `)
+                .else()
+                .code(() => cxt.error())
+                .endIf();
+        }
+    },
+
+    // schema: { binary: 64 } will match Buffer.alloc(64)
+    // schema: { binary: true } will match Buffer.alloc(<any>)
+    binary: {
+        keyword: 'binary',
+        // schemaType: ['boolean', 'number'],
+        /**
+         * 
+         * @param {KeywordCxt} cxt 
+         * 
+         */
+        code(cxt) {
+            const d = cxt.data;
+            const s = cxt.schemaValue;
+            cxt.gen
+                .if(CG`
+                    Buffer.isBuffer(${d}) && (
+                        typeof ${s} === 'number' ? 
+                        ${d}.length === ${s} :
+                        ${s}
+                    )
+                `)
+                .elseIf(CG`
+                    ${d}._bsontype === 'Binary' && (
+                        typeof ${s} === 'number' ? 
+                        ${d}.length() === ${s} :
+                        ${s}
+                    )
+                `)
+                .else()
+                .code(() => cxt.error())
+                .endIf();
+        }
+    },
+
+    wrapper: {
+        keyword: 'wrapper',
+        errors: false,
+        modifying: true,
+        /**
+         * 
+         * @param {KeywordCxt} cxt 
+         * 
+         */
+        code(cxt) {
+            if (cxt.it.dataLevel <= 0) return;
```

Comment:
I think this check was meant to make sure we did not use `wrapper` on a root schema, and always have a `type: 'object'` above that uses a wrapper on one of its properties, so that we can modify it to the wrapped value. But I think that returning is not what we should do - instead we should actively fail the schema compilation. Perhaps just throw an error from here.

Reply:
Not sure that I follow, wouldn't it fail on addSchema? 

something like:
```
     Error: schema is invalid: data/properties/wrapper must be object,boolean
      at Ajv.validateSchema (node_modules/ajv/dist/core.js:263:23)
      at Ajv._addSchema (node_modules/ajv/dist/core.js:457:18)
      at Ajv.addSchema (node_modules/ajv/dist/core.js:231:34)
 ```
If it is indeed the case then this check is not necessary here also.

Reply:
This checks that there is a parent schema, to avoid having a top level schema that is only `schema = { wrapper: SensitiveString }` because then we won't be able to replace the property in the parent.

Reply:
so I am thinking of changing the code to:

```
        code(cxt) {
            const dl = cxt.it.dataLevel;
            cxt.gen
                .if(CG `${dl} <= 0`)
                .code(() => cxt.error())
                .endIf();
            const d = cxt.data;
         ...
```

and for test:
```
const test_schema_warpper_root = {
        $id: 'test',
        wrapper: SensitiveString,
}


mocha.describe('Test Schema Keywords', function() {

    mocha.before('Adding Schema And Keywords', async function() {
        ajv.addSchema(test_schema_warpper_root);
        ajv.addKeyword(schema_keywords.KEYWORDS.wrapper);
    });

    mocha.it('test', async function() {
        const validator = ajv.getSchema('test#/');
        assert.strictEqual(validator(new SensitiveString('test')), false);
    });
}
```

WDYT?

Reply:
Done.